diff -rcN src/binutils-org/bfd/config.bfd src/binutils/bfd/config.bfd
*** src/binutils-org/bfd/config.bfd	2019-06-28 17:04:58.398645609 +0200
--- src/binutils/bfd/config.bfd	2019-06-28 16:53:02.027064030 +0200
***************
*** 224,229 ****
--- 224,234 ----
  
  # START OF targmatch.h
  #ifdef BFD64
+   x86_64-*-m3)
+     targ_defvec=x86_64_elf64_vec
+     targ_selvecs="i386_elf32_vec"
+     want64=true
+     ;;
    aarch64-*-darwin*)
      targ_defvec=aarch64_mach_o_vec
      targ_selvecs="arm_mach_o_vec mach_o_le_vec mach_o_be_vec mach_o_fat_vec"
diff -rcN src/binutils-org/bfd/elfxx-x86.c src/binutils/bfd/elfxx-x86.c
*** src/binutils-org/bfd/elfxx-x86.c	2019-06-28 17:03:36.118081098 +0200
--- src/binutils/bfd/elfxx-x86.c	2019-06-28 17:03:38.714765581 +0200
***************
*** 27,35 ****
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF32_DYNAMIC_INTERPRETER "/usr/lib/libc.so.1"
! #define ELF64_DYNAMIC_INTERPRETER "/lib/ld64.so.1"
! #define ELFX32_DYNAMIC_INTERPRETER "/lib/ldx32.so.1"
  
  bfd_boolean
  _bfd_x86_elf_mkobject (bfd *abfd)
--- 27,35 ----
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF32_DYNAMIC_INTERPRETER "/bin/dynlink"
! #define ELF64_DYNAMIC_INTERPRETER "/bin/dynlink"
! #define ELFX32_DYNAMIC_INTERPRETER "/bin/dynlink"
  
  bfd_boolean
  _bfd_x86_elf_mkobject (bfd *abfd)
diff -rcN src/binutils-org/config.sub src/binutils/config.sub
*** src/binutils-org/config.sub	2019-06-28 17:03:03.027854077 +0200
--- src/binutils/config.sub	2019-06-28 17:08:28.450086747 +0200
***************
*** 1339,1345 ****
  	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
  	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
  	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
! 	     | sym* | kopensolaris* | plan9* \
  	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
  	     | aos* | aros* | cloudabi* | sortix* \
  	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
--- 1339,1345 ----
  	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
  	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
  	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
! 	     | sym* | kopensolaris* | plan9* | m3* \
  	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
  	     | aos* | aros* | cloudabi* | sortix* \
  	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
diff -rcN src/binutils-org/gas/configure.tgt src/binutils/gas/configure.tgt
*** src/binutils-org/gas/configure.tgt	2019-06-28 17:03:03.321189419 +0200
--- src/binutils/gas/configure.tgt	2019-06-28 17:14:09.729094913 +0200
***************
*** 121,126 ****
--- 121,128 ----
  generic_target=${cpu_type}-$vendor-$os
  # Note: This table is alpha-sorted, please try to keep it that way.
  case ${generic_target} in
+   i*86-*-m3*)     fmt=elf ;;
+   x86_64-*-m3*)     fmt=elf ;;
    aarch64*-*-elf*)			fmt=elf;;
    aarch64*-*-fuchsia*)			fmt=elf;;
    aarch64*-*-linux*)			fmt=elf em=linux
diff -rcN src/binutils-org/ld/configure.tgt src/binutils/ld/configure.tgt
*** src/binutils-org/ld/configure.tgt	2019-06-28 17:03:03.067854350 +0200
--- src/binutils/ld/configure.tgt	2019-06-28 17:10:49.124385178 +0200
***************
*** 251,256 ****
--- 251,257 ----
  			targ_extra_emuls="elf_iamcu"
  			targ64_extra_emuls="elf_x86_64 elf32_x86_64 elf_l1om elf_k1om"
  			targ64_extra_libpath="elf_x86_64 elf32_x86_64" ;;
+ x86_64-*-m3) targ_emul=m3_x86_64 ;;
  x86_64-*-linux-gnux32)	targ_emul=elf32_x86_64
  			targ_extra_emuls="elf_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om"
  			targ_extra_libpath="elf_i386 elf_iamcu elf_x86_64 elf_l1om elf_k1om"
diff -rcN src/binutils-org/ld/emulparams/m3_x86_64.sh src/binutils/ld/emulparams/m3_x86_64.sh
*** src/binutils-org/ld/emulparams/m3_x86_64.sh	1970-01-01 01:00:00.000000000 +0100
--- src/binutils/ld/emulparams/m3_x86_64.sh	2019-06-28 16:53:02.030397386 +0200
***************
*** 0 ****
--- 1,17 ----
+ SCRIPT_NAME=elf
+ ELFSIZE=64
+ OUTPUT_FORMAT=elf64-x86-64
+ TEXT_START_ADDR=0x00001000
+ MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+ COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ ARCH="i386:x86-64"
+ MACHINE=
+ NOP=0x90909090
+ TEMPLATE_NAME=elf32
+ GENERATE_SHLIB_SCRIPT=yes
+ GENERATE_PIE_SCRIPT=yes
+ NO_SMALL_DATA=yes
+ SEPARATE_GOTPLT=24
+ DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ SHLIB_DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ 
diff -rcN src/binutils-org/ld/Makefile.in src/binutils/ld/Makefile.in
*** src/binutils-org/ld/Makefile.in	2019-06-28 17:13:18.852079190 +0200
--- src/binutils/ld/Makefile.in	2019-06-28 16:53:02.030397386 +0200
***************
*** 645,650 ****
--- 645,651 ----
  	eaixrs6.c \
  	ealpha.c \
  	ealphavms.c \
+ 	em3_x86_64.c \
  	earcv2elf.c \
  	earcv2elfx.c \
  	earcelf.c \
***************
*** 2265,2270 ****
--- 2266,2274 ----
    $(srcdir)/emultempl/needrelax.em \
    $(srcdir)/scripttempl/alphavms.sc ${GEN_DEPENDS}
  
+ em3_x86_64.c: $(srcdir)/emulparams/m3_x86_64.sh \
+   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+ 
  earcv2elf.c: $(srcdir)/emulparams/arcv2elf.sh \
    $(ELF_DEPS) $(srcdir)/scripttempl/elfarcv2.sc ${GEN_DEPENDS}
  
