diff -rcN src/gdb-org/bfd/config.bfd src/gdb/bfd/config.bfd
*** src/gdb-org/bfd/config.bfd	2019-07-15 13:27:11.763854046 +0200
--- src/gdb/bfd/config.bfd	2019-07-15 13:31:26.810532432 +0200
***************
*** 224,229 ****
--- 224,234 ----
  
  # START OF targmatch.h
  #ifdef BFD64
+   x86_64-*-m3)
+     targ_defvec=x86_64_elf64_vec
+     targ_selvecs="i386_elf32_vec"
+     want64=true
+     ;;
    aarch64-*-darwin*)
      targ_defvec=aarch64_mach_o_vec
      targ_selvecs="arm_mach_o_vec mach_o_le_vec mach_o_be_vec mach_o_fat_vec"
diff -rcN src/gdb-org/config.sub src/gdb/config.sub
*** src/gdb-org/config.sub	2019-07-15 13:27:11.730520712 +0200
--- src/gdb/config.sub	2019-07-15 13:29:34.707193914 +0200
***************
*** 1339,1345 ****
  	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
  	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
  	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
! 	     | sym* | kopensolaris* | plan9* \
  	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
  	     | aos* | aros* | cloudabi* | sortix* \
  	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
--- 1339,1345 ----
  	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
  	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
  	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
! 	     | sym* | kopensolaris* | plan9* | m3* \
  	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
  	     | aos* | aros* | cloudabi* | sortix* \
  	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
